src/tools/create-http-patch.js

```diff
@@ -0,0 +1,76 @@
+const _ = require('lodash');
+const constants = require('../constants');
+
+const createHttpPatch = (event) => {
+  const createLogger = require('./create-logger');
+  const logBuffer = [];
+  const logger = createLogger(event, {logBuffer});
+
+  const httpPatch = (object) => {
+    const originalRequest = object.request;
+
+    // Avoids multiple patching and memory leaks (mostly when running tests locally)
+    if (object.patchedByZapier) {
+      return;
+    }
+
+    object.patchedByZapier = true;
+
+    // Proxy the request method
+    object.request = (options, callback) => {
+      const requestUrl = options.url || options.href || (`${options.protocol || 'https://'}${options.host}${options.path}`);
+      const logger_url = process.env.LOGGING_ENDPOINT || constants.DEFAULT_LOGGING_HTTP_ENDPOINT;
+
+      // Ignore logger requests
+      if (requestUrl.indexOf(logger_url) !== -1) {
+        return originalRequest(options, callback);
+      }
+
+      // Ignore requests made via the request client
+      if (_.get(options.headers, 'user-agent', []).indexOf('Zapier') !== -1) {
+        return originalRequest(options, callback);
+      }
+
+      // Proxy the callback to get the response
+      const newCallback = function(response) {
+        const requestHeaders = _.map(Object.keys(options.headers || {}), (header) => `${header}: ${options.headers[header]}`);
+        const responseHeaders = _.map(Object.keys(response.headers || {}), (header) => `${header}: ${response.headers[header]}`);
+        const responseBody = [];
+
+        const logResponse = () => {
+          // Prepare data for GL
+          const logData = {
+            log_type: 'http',
+            request_type: 'devplatform-outbound',
+            request_url: requestUrl,
+            request_method: options.method || 'GET',
+            request_headers: requestHeaders.join('\n'),
+            request_data: options.body || '',
+            request_via_client: false,
+            response_status_code: response.statusCode,
+            response_headers: responseHeaders.join('\n'),
+            response_content: responseBody.join(''),
+          };
+
+          logger(`${logData.response_status_code} ${logData.request_method} ${logData.request_url}`, logData);
+        };
+
+        response.setEncoding('utf8');
+        response.on('data', (chunk) => responseBody.push(chunk));
+        response.on('end', logResponse);
+        response.on('error', logResponse);
```

Comment:
I don't think we can safely attach all of these. I wrote [a gist](https://gist.github.com/bcooksey/2dc38ffd18f0eda87fb1f79d4c56fed2) to demonstrate some behaviors and found this:

1) `setEncoding()` swaps the incoming values to the data listener from buffers to strings, which is probably not a nice thing to do to devs
2) `setEncoding()` will, obviously, encode the data, so if binary data being transferred, devs would have to decode it to get back to what they want
3) Adding a data handler [swaps the stream mode](https://nodejs.org/dist/latest-v6.x/docs/api/stream.html#stream_two_modes) from paused to flowing. This can cause `.read()` to return no data. As far as how much we should care, seems like an app will end up doing a `.pipe()` or something and get it into flow mode anyway, but it is a nasty edge to force on folks. If we knew consuming data via `.read()` was rare, I'd be fine ignoring this point.

Reply:
Wow, nice catches! I'll see if I can tap into that without changing data (will use your gist).

If I can't, I'll remove that and we'll log without the response content (still better than nothing). If it becomes something we _really_ want, we can then try to look into ways of tapping into it more safely, maybe.

Reply:
You may also want to look into how NewRelic wraps `http.request` and friends. Pretty sure they've made sure not to break it: https://github.com/newrelic/node-newrelic/blob/master/lib/instrumentation/core/http.js#L579-L606

Reply:
Thanks @FokkeZB. After looking through it and trying it out for a while, it seems their wrapper libraries/methods are very complex and generic (and the request/response parts use the agent and eventemitters, which, while _probably better_, seem like a major overkill for what we're going for here).

I was able to confirm no app in CLI is currently calling anything `.read()`, so I think we're safe on @bcooksey 's third concern. I ditched the `setEncoding` and that seems to still be working fine (worst case scenario we might have some encoding issues on GL, and we can try to tweak that if so, later).

Let me know if that works fine.